src/edgeiter.jl

```diff
@@ -0,0 +1,221 @@
+#                 state.fin = true
+#                 return (edge, state)
+#             end
+#         end
+#     end
+#     state.fin = _isfin(eit,state)
+#     return (edge, state)
+# end
+#
+#
+# #
+# #
+# #     d = 0
+# #     println("top of next: eit = $eit, state = $state")
+# #     while state.di <= length(eit.adj[state.s])      # while the index to dest vertex is still within the src vertex vector
+# #         d  = eit.adj[state.s][state.di]             # set d to the dest vertex
+# #         !eit.directed && d >= state.s && break      # if the dest vertex is >= to the source, stop the loop
+# #         state.di += 1                               # else increase the dest vertex index
+# #     end
+# #     while state.di > length(eit.adj[state.s])       # while the index to the dest vertex is greater than the src vertex vector
+# #         state.di = 1                                # reset the dest vertex indicator
+# #         state.s += 1                                # go to the next source vertex vector
+# #         while state.di <= length(eit.adj[state.s])  # do it all again.
+# #             d = eit.adj[state.s][state.di]
+# #             !eit.directed && d >= state.s && break
+# #             state.di += 1
+# #         end
+# #     end
+# #     e = !eit.directed && d < state.s ? Edge(d, state.s) : Edge(state.s, d)
+# #     state.di += 1
+# #     state.it += 1
+# #     return (e, state)
+# # end
+#
+# length(eit::EdgeIter) = eit.m
+#
+# # note: e in edges(g) will not necessarily be the same as
+# # e in collect(edges(g)) for undirected graphs. The former will be true
+# # when s > d; the latter will not.
+# function in(e::Edge, eit::EdgeIter)
+#     if !is_ordered(e)
+#         e = reverse(e)
+#     end
+#     s, t = e
+#     n = length(eit.adj)
+#     !(1 <=  s <= n) && return false
+#     !(1 <=  t <= n) && return false
+#     return t in eit.adj[s]
+# end
+
+# function getindex(eit::EdgeIter, n::Int)
+#     1 <= n <= eit.m || throw(BoundsError())
+#     offsetsum = 0
+#     i = 1
+#     e = Edge(0,0)
+#     ordered = false
+#     while !ordered
+#         info("top of loop: n = $n, offsetsum = $offsetsum, i = $i, e = $e")
+#         while (offsetsum < n)
+#             offsetsum += length(eit.adj[i])
+#             i += 1
+#         end
+#         i -= 1
+#         info("out of loop: n = $n, offsetsum = $offsetsum, i = $i, e = $e")
+#         offsetsum -= length(eit.adj[i])
+#         d = eit.adj[i][n - offsetsum]
+#         e = Edge(i,d)
+#         info("pre-order check: n = $n, offsetsum = $offsetsum, i = $i, e = $e")
+#         ordered = is_ordered(e)
+#         if !ordered
+#             n += 1
+#         end
+#     end
+#     return e
+# end
+
+
+# getindex(eit::EdgeIter, n::UnitRange{Int64}) = [getindex(eit, x) for x in n]
+
+#TODO implement using a loop and ∈, so that no memory is allocated
+function _isequal(e1::EdgeIter, e2)
+    for e in e2
+        s, d = e
+        found = length(searchsorted(e1.adj[s], d)) > 0
```

Comment:
By running `@edit zip([1,2,3],[4,5,6])` I discovered that zip returns an iterator with O(1) memory cost.

```julia
julia> function is_eq(xs,ys)
       for (i,j) in zip(xs,ys)
           i==j||return false
       end
       return true
       end
is_eq (generic function with 1 method)

julia> is_eq([1,2,3],[1,2,3])
true

julia> is_eq([1,2,3],[1,2,4])
false
```